Update timezone.md (#2138)
tc39/proposal-temporal: docs/timezone.md

/* WRONG */ tz = new Temporal.TimeZone('local'); // => throws, not a time zone
```

> #### Difference between IANA time zones and simple UTC offsets

Reviewer: I'd suggest calling these "numeric UTC offsets", not "simple".

> The returned time zone object behaves slightly differently depending on whether an IANA time zone name (e.g., `Europe/Berlin`) is given, or a UTC offset (e.g., `+01:00`).

Reviewer: ```suggestion
The returned time zone object behaves slightly differently depending on whether an IANA time zone name (e.g., `Europe/Berlin`) is given, or a numeric UTC offset (e.g., `+01:00`).
```

**Returns:** The UTC offset at the given time, in nanoseconds.

Since the UTC offset can change throughout the year in time zones that employ DST as well as because of special political decisions, this method queries the UTC offset at a particular time.

> Note that only `Temporal.TimeZone` objects constructed from an IANA time zone name may have UTC offset transitions; those constructed from a UTC offset are fixed.

Reviewer: I agree with your changes. While reading the original text, I think the original text could be clearer. So on top of your changes, would it be worth clarifying the text further? Maybe something like this to replace this and the sentence below it?

```suggestion
Note that `Temporal.TimeZone` objects constructed from an IANA time zone name may change offsets depending on the instant.
However, other time zones (some IANA time zones like `Etc\GMT+5` and all time zones constructed from a numeric UTC offsets) have fixed offsets that never change, regardless of the instant.
```

This method is similar to `timeZone.getOffsetNanosecondsFor()`, but returns the offset formatted as a string, with sign, hours, and minutes.

> If `timeZone` is a time zone constructed from a fixed UTC offset, the return value of this method is effectively the same as `timeZone.id`.

Reviewer: ```suggestion
If `timeZone` is a time zone constructed from a numeric UTC offset, the return value of this method is effectively the same as `timeZone.id`.
```

If `instant` is not a `Temporal.Instant` object, then it will be converted to one as if it were passed to `Temporal.Instant.from()`.

> When subclassing `Temporal.TimeZone`, this method doesn't need to be overridden, because the default implementation creates a `Temporal.PlainDateTime` from `instant` using a UTC offset which is the result of calling `timeZone.getOffsetNanosecondsFor()`.

Reviewer: This is a subtle grammar change, but I think it's correct with the extra comma. Nice find.

- `'later'`: The later of two possible times.
- `'reject'`: Throw a `RangeError` instead.

When interoperating with existing code or services, `'compatible'` mode matches the behavior of legacy `Date` as well as libraries like Moment.js, Luxon, and date-fns.
This mode also matches the behavior of cross-platform standards like [RFC 5545 (iCalendar)](https://tools.ietf.org/html/rfc5545).

> During "skipped" clock time, like the hour after DST starts in the spring, or when a country permanently changed its UTC offset to one closer to positive infinity, this method interprets invalid times using the pre-transition time zone offset if `'compatible'` or `'later'` is used, or the post-transition time zone offset if `'earlier'` is used.

Reviewer: > permanently changed its UTC offset to one closer to positive infinity

Could this be said more simply?  Perhaps "larger" or "smaller" ?

Reviewer: Yep, I agree with @ptomato-- seems best to leave as-is. Changing the capitalization to "spring" is correct though!  @h-h-h-h you have a great eye for grammar mistakes and typos!

If `dateTime` is not a `Temporal.PlainDateTime` object, then it will be converted to one as if it were passed to `Temporal.PlainDateTime.from()`.

> Normally there is only one possible exact time corresponding to a wall-clock time, but around a daylight saving or other political change, a wall-clock time may not exist, or the same wall-clock hour sequence may exist twice in a row (for DST usually just one double-hour).

Reviewer: What's an "hour sequence" ?  Would be good to understand why "time" isn't OK here. I'd prefer to keep simpler terms when it makes sense.

Author: The IANA time zone ["Antarctica/Troll"](https://en.wikipedia.org/wiki/Time_in_Norway#IANA_time_zone_database) has a DST switch of 2 hours, meaning the sequence of local hours can be as follows: <code>22, 23, ***00, 01, 00, 01,*** 02, 03</code>. `00, 01` is an hour sequence and that *sequence* happens twice. The same could possibly happen in the context of permanent UTC offset changes. It shouldn't be thought that it'll definitely always just be one double-hour.

I clarified it a bit. See next commit.

Reviewer 2: My opinion is it'll be really confusing if we use the rare case of a 2-hour DST transition as the example here.

I think the original wording ("same wall-clock time may exist twice in a row") should cover that case as well, or maybe remove "in a row" if you think it too heavily implies "within one 60-minute time span". The same wall-clock time existing twice, seems like it can describe 1-hour DST transitions, 2-hour DST transitions, or time zone boundary changes, as long as they are in the "fall back" direction.

Reviewer: I agree with @ptomato. We should try to avoid adding complexity to accommodate rare cases unless it's likely that programmer bugs will result from ignoring those cases. 

What about wording like below, which is shorter, simpler, and makes no assumptions that repeats are one hour long?

> Normally there is only one possible exact time corresponding to a wall-clock time, but around a daylight saving or other offset change, a wall-clock time may not exist, or the same wall-clock time may exist twice.

Author: I applied @justingrant's proposal with exception of the word "exactly" in "exactly one possible exact time" (keeping original "only").

During "skipped" clock time like the hour after DST starts in the spring or certain political changes, `Temporal.TimeZone.prototype.getInstantFor()` returns a `Temporal.Instant` (by default interpreting the `Temporal.PlainDateTime` using the pre-transition time zone offset), while this method returns zero results during those skipped periods.

### timeZone.**getNextTransition**(_startingPoint_: Temporal.Instant | string) : Temporal.Instant

**Parameters:**

- `startingPoint` (`Temporal.Instant` or value convertible to one): Time after which to find the next UTC offset transition.

**Returns:** A `Temporal.Instant` object representing the next UTC offset transition in this time zone, or `null` if no transitions later than `startingPoint` could be found.

This method is used to calculate a possible future UTC offset transition after `startingPoint` for this time zone. This can be because of DST or other political changes like a country having permanently changed its offset.

> Note that if the time zone was constructed from a UTC offset, there will be no offset transitions.

Reviewer: ```suggestion
Note that if the time zone was constructed from a UTC offset, there will be no offset transitions.
Also note that some IANA time zones (e.g., `Etc/GMT+5`) have no offset transitions either.
```

**Parameters:**

- `startingPoint` (`Temporal.Instant` or value convertible to one): Time before which to find the previous UTC offset transition.

**Returns:** A `Temporal.Instant` object representing the previous UTC offset transition in this time zone, or `null` if no transitions earlier than `startingPoint` could be found.

This method is used to calculate a possible past UTC offset transition after `startingPoint` for this time zone. This can be because of DST or other political changes like a country having permanently changed its offset.

> Note that if the time zone was constructed from a UTC offset, there will be no offset transitions.

Reviewer: ```suggestion
Note that if the time zone was constructed from a UTC offset, there will be no offset transitions.
Also note that some IANA time zones (e.g., `Etc/GMT+5`) have no offset transitions either.
```
